torch/distributed/_sharding_spec/api.py

```diff
@@ -0,0 +1,98 @@
+from abc import ABC
+import torch
+from typing import List, Union
+
+from ._internals import is_valid_device
+
+Device = Union[torch.device, int, str]
+
+class PlacementSpec(ABC):
+    """
+    Base class representing the placement of an entity.
+    """
+    pass
+
+class DevicePlacementSpec(PlacementSpec):
+    """
+    Associates placement of an entity with a single device. The device can be a
+    local device or a remote device specified by one of the following remote
+    formats:
+
+        1. "rank:<rank>/<device>" (ex: "rank:0/cuda:0").
+        2. "<worker_name>/<device>" (ex: "trainer0/cuda:0").
+
+    Args:
+        device(str, :class:`torch.device`): The device to place the entity on.
+    """
+    def __init__(self, device: Device):
+        super(DevicePlacementSpec, self).__init__()
+        if not is_valid_device(device):
+            raise ValueError(f'{device} is not a valid device')
+        self._device = device
+
+    @property
+    def device(self) -> Device:
+        """
+        Retrieves the device for placement.
+        """
+        return self._device
+
+class ChunkShardingSpec(PlacementSpec):
+    """
+    This is type of PlacementSpec that defines the placement as being sharded
+    across multiple devices. In particular, it represents sharding a Tensor
+    along a single dimension into equal chunks (similar to :meth:`torch.chunk`).
+
+    Args:
+        dim (int or str):
+            The dimension to shard on, could be an integer representing the
+            dimension or a string in case of named tensors where dimensions are
+            named.
+        placement(List[Device] or List[PlacementSpec]):
```

Comment:
Should `List[PlacementSpec]` be `List[DevicePlacementSpec]`? Otherwise, it means that a `ChunkShardingSpec` can be represented by `List[ChunkShardingSpec]`, which means that the sharding here can be hierarchical.

Reply:
I kept it `List[PlacementSpec]` for extra flexibility. For example you could subclass PlacementSpec to specify your own custom placement (ex: some database like rocksdb).

Reply:
Can you add a comment about this? So powerful users know that they can subclass `PlacementSpec` for this purpose.